### PR TITLE
[ROCm][Perf] Add bpreshuffled blockscaled fp8 GEMM

### PR DESCRIPTION
**Pull request by simondanielsson**
## Purpose

Activated when shapes allow for it and configs are tuned.

**Implications (DSv3 1k/1K):** 
- TP8+DPA: +4-8% QPS
- TP8+EP: +0-4% QPS

## Test Plan

Bench serve & accuracy validation with DSv3 on (1) TP8+DPA (2) TP8+EP on 8xMI350.

(note need to run with `VLLM_ROCM_USE_AITER_FP8BMM=0` until https://github.com/vllm-project/vllm/issues/51957 is resolved)
```bash
VLLM_ROCM_USE_AITER=1 \
VLLM_ROCM_USE_AITER_FP8BMM=0 \
vllm serve deepseek-ai/DeepSeek-V3-0324 \
    --data-parallel-size 8 \
    --tensor-parallel-size 1 \
    --gpu-memory-utilization 0.8 \
```

Sweep concurrencies 1-128, 1k/1k
```shell
vllm bench serve \
    --backend vllm \
    --model deepseek-ai/DeepSeek-V3-0324 \
    --dataset-name random \
    --random-input-len 1024 \
    --random-output-len 1024 \
    --host localhost \
    --port 8000 \
    --ignore-eos \
    --ready-check-timeout-sec 6000 \
    --seed 5678 \
    --max-concurrency 128 \
    --num-prompts 1280 --num-warmups 256'
```

```shell
pip install --quiet 'lm_eval[api]' && \
  lm_eval \
    --model local-completions \
    --model_args model=deepseek-ai/DeepSeek-V3-0324,base_url=http://localhost:8000/v1/completions,tokenized_requests=False,trust_remote_code=True,num_concurrent=256 \
    --tasks gsm8k \
    --num_fewshot 5 \
    --output_path /tmp/lm_eval_gsm8k
```

### Profiling

This branch: Using DPA, the MLA decode o-proj is using preshuffled weights. No other GEMM is changed. 21us in trace.

image

Main branch: ck-tile kernel for o-proj. 52us in trace, so **more than 2x slower that of the bpreshuffled version!**

image

### TP8+DPA

| Concurrency | Branch      | TTFT (P50 ms) | TPOT (P50 ms) | QPS    | QPS change (%) |
|-------------|-------------|---------------|---------------|--------|----------------|
| 1           | This branch | 156.9816      | 18.2976       | 0.0529 | +7.30          |
| 1           | Nightly     | 158.3445      | 19.7720       | 0.0493 |               |
| 2           | This branch | 161.0075      | 18.5311       | 0.1049 | +8.14          |
| 2           | Nightly     | 198.6506      | 19.9552       | 0.0970 |   |
| 4           | This branch | 220.7625      | 18.7688       | 0.2054 | +7.48          |
| 4           | Nightly     | 284.2663      | 20.1512       | 0.1911 |               |
| 8           | This branch | 557.7459      | 22.6129       | 0.3479 | +7.61          |
| 8           | Nightly     | 317.1527      | 23.8614       | 0.3233 |               |
| 16          | This branch | 581.6531      | 23.5224       | 0.6528 | +5.96          |
| 16          | Nightly     | 609.4239      | 24.8873       | 0.6161 |               |
| 32          | This branch | 767.6995      | 23.7180       | 1.2794 | +4.61          |
| 32          | Nightly     | 787.4866      | 24.8746       | 1.2230 |               |
| 64          | This branch | 1449.0858     | 24.3342       | 2.4326 | +5.50          |
| 64          | Nightly     | 1394.4616     | 25.7091       | 2.3057 |               |
| 128         | This branch | 2018.3187     | 25.9321       | 4.4837 | +4.90          |
| 128         | Nightly     | 2201.5545     | 27.2313       | 4.2742 |               |

### TP8+EP

| Concurrency | Branch      | TTFT (P50 ms) | TPOT (P50 ms) | QPS    | QPS change (%) |
|-------------|-------------|---------------|---------------|--------|----------------|
| 1           | This branch | 108.3783      | 12.6642       | 0.0765 | -0.39          |
| 1           | Nightly     | 113.4111      | 12.6244       | 0.0768 |                |
| 2           | This branch | 172.7154      | 13.6376       | 0.1416 | +0.14          |
| 2           | Nightly     | 178.4821      | 13.6537       | 0.1414 |                |
| 4           | This branch | 280.6493      | 13.3944       | 0.2861 | +0.42          |
| 4           | Nightly     | 287.1797      | 13.446        | 0.2849 |                |
| 8           | This branch | 424.7564      | 13.5723       | 0.5592 | +0.81          |
| 8           | Nightly     | 427.5746      | 13.6773       | 0.5547 |                |
| 16          | This branch | 479.1289      | 14.4931       | 1.0434 | +0.69          |
| 16          | Nightly     | 484.925       | 14.6155       | 1.0363 |                |
| 32          | This branch | 817.9601      | 15.6236       | 1.9015 | +3.95          |
| 32          | Nightly     | 821.0856      | 15.8782       | 1.8292 |                |
| 64          | This branch | 938.6096      | 21.6289       | 2.7676 | +9.01          |
| 64          | Nightly     | 932.3192      | 23.9109       | 2.5388 |                |
| 128         | This branch | 1064.6245     | 27.3408       | 4.4065 | +4.04          |
| 128         | Nightly     | 1079.7563     | 28.087        | 4.2355 |                |

### GSM8k

This branch (TP8+DPA)

|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.9477|±  |0.0061|
|     |       |strict-match    |     5|exact_match|↑  |0.9484|±  |0.006

Nightly

|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.9439|±  |0.0063|
|     |       |strict-match    |     5|exact_match|↑  |0.9431|±  |0.0064|

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
</details>

### CODE REVIEW

File: vllm/_aiter_ops.py [hunk 1]
```diff
@@ -110,6 +110,21 @@ def _check_kernel_tuned(N: int, K: int, q_dtype_w: torch.dtype, csv_path: str) -
     return any((N, K, M) in configs for M in l_m)
 
 
+@functools.cache
```
Comment (Rohan138) [inline vllm/_aiter_ops.py]: There's a `_load_gemm_tuned_configs` helper, can we use that instead?

Comment (simondanielsson) [reply vllm/_aiter_ops.py]: We can't immediately use `_load_gemm_tuned_configs` because we're filtering on other fields that the q_dtype_w here and also we don't necessarily need all of N/M/K. 

However we could generalize it. WDYT @Rohan138?
```python
@functools.cache
def _load_gemm_tuned_configs(
    csv_path: str,
    filters: tuple[tuple[str, object], ...],
    key_cols: tuple[str, ...] = ("N", "K", "M"),
) -> set[tuple[int, ...]]:
    try:
        df = pd.read_csv(csv_path).drop_duplicates()
        for col, val in filters:
            if col not in df.columns:
                continue
            if isinstance(val, int):
                df = df[df[col].astype(int) == val]
            else:
                df = df[df[col].astype(str) == str(val)]
        return set(zip(*(df[c].astype(int) for c in key_cols)))
    except Exception:
        return set()

# and then just call 
        configs = _load_gemm_tuned_configs(csv_path, (("q_dtype_w", q_dtype_w),))

# or for the code added in this PR
        return (n, k) in _load_gemm_tuned_configs(
            csv_path, (("gfx", gfx), ("cu_num", cu_num)), key_cols=("N", "K")
        )

```

Comment (simondanielsson) [reply vllm/_aiter_ops.py]: IMO I prefer the two different versions as a unified function `_load_gemm_tuned_configs` might become overly complex easily. Happy to hear your thuoghs though @Rohan138

**Comment by simondanielsson** [comment]
@Rohan138 Happy to get a re-review on this one whenever you have time, thanks 🙏

File: vllm/_aiter_ops.py (hunk 1 above)
Comment (Rohan138) [reply vllm/_aiter_ops.py]: honestly I slightly prefer the generalization for now ... we'd eventually like to stop checking them here and have AITER (through e.g. tgemm) just choose the tuned configs or fall back as needed

Comment (simondanielsson) [reply vllm/_aiter_ops.py]: @Rohan138 Updated to the generlzied version - PTAL! Thanks

Agree, and it will be easier to remove this logic this way once (if) aiter itself dispatches based on tuned/not tuned cfgs

File: vllm/model_executor/kernels/linear/scaled_mm/aiter.py [hunk 2]
```diff
@@ -395,6 +404,28 @@ def process_weights_after_loading(self, layer: torch.nn.Module) -> None:
         if ws is not None and ws.dtype == torch.float8_e8m0fnu:
             replace_parameter(layer, attr, _upcast_e8m0_to_fp32(ws).contiguous())
 
+        if not self._is_bpreshuffled:
```
Comment (BadrBasowid) [inline vllm/model_executor/kernels/linear/scaled_mm/aiter.py]: An early return in this method would mean that `AiterFp8BlockScaledMMKernel.apply_weights` could be called with the wrong weight layout, right? Perhaps this can be moved to`can_implement` method?

**Review by BadrBasowid — COMMENTED** [review]
Small nit

File: vllm/model_executor/kernels/linear/scaled_mm/aiter.py (hunk 2 above)
Comment (tjtanaa) [reply vllm/model_executor/kernels/linear/scaled_mm/aiter.py]: right. And I think it is better to create a `AiterPreshuffledFp8BlockScaledMMKernel` same as https://github.com/simondanielsson/vllm/blob/7752ecb8ac83fc27b687de672711efc5b8d94d32/vllm/model_executor/kernels/linear/scaled_mm/aiter.py#L132

Comment (BadrBasowid) [reply vllm/model_executor/kernels/linear/scaled_mm/aiter.py]: Oh, I just saw the `if not self._is_bpreshuffled` branch in `apply_weights`, so my previous concern was invalid, but i think @tjtanaa raises a good suggestion, since it's not quite obvious what's being dispatched

Comment (BadrBasowid) [reply vllm/model_executor/kernels/linear/scaled_mm/aiter.py]: Creating a separate class would also be better for logging, since it clearly prints which kernel class has been selected

Comment (simondanielsson) [reply vllm/model_executor/kernels/linear/scaled_mm/aiter.py]: @tjtanaa  @BadrBasowid  Very good suggestion, this is much nicer. Fixed - PTAL 🙏 

Now from logs:
```
(Worker_DP0 pid=5537) INFO 08-20 16:23:06 [__init__.py:690] Selected AiterPreshuffledFp8BlockScaledMMKernel for Fp8LinearMethod
```
and e.g. o-proj still uses preshuffled gemm:

image

**Comment by simondanielsson** [comment]
@tjtanaa Ready for review when you have time, thanks 🙏

**Review by BadrBasowid — APPROVED** [review]
LGTM !

**Review by tjtanaa — APPROVED** [review]
LGTM.

**Comment by AndreasKaratzas** [comment]
@simondanielsson I cancelled this build cause like the other 2 it creates a huge workload for our CI. When you relaunch CI, please wait for each PR to finish and then launch CI for you next PR.

**Comment by simondanielsson** [comment]
@AndreasKaratzas Same on this PR. Only failing tests are the same as in nightly. Can you check please? 🙏 Thanks

**Comment by AndreasKaratzas** [comment]
Indeed they are from nightly.

**Comment by mawong-amd** [comment]
Was this PR tested on DeepSeek V3/V4 TP8? I'm seeing garbage output on the former, and it refuses to initialize on the latter (it hits the not implemented `apply_block_scaled_mm`).
Note: Testing on `main` now is slightly complicated because https://github.com/vllm-project/vllm/pull/56228 changed the signature of the sparse-attention indexer call (a separate breakage), but the results stand after patching that.

**Comment by simondanielsson** [comment]
> Was this PR tested on DeepSeek V3/V4 TP8? I'm seeing garbage output on the former, and it refuses to initialize on the latter (it hits the not implemented `apply_block_scaled_mm`). Note: Testing on `main` now is slightly complicated because #56228 changed the signature of the sparse-attention indexer call (a separate breakage), but the results stand after patching that.

@mawong-amd Hi, yes tested on V3 TP8 (note: vanilla dense v3). Let me also test again
